File: examples/haskell-http/Transport.hs
{-# LANGUAGE ForeignFunctionInterface, OverloadedStrings #-}
module Transport (withTransport) where

import Control.Exception (bracket, onException)
import Control.Monad (unless)
import Data.Aeson
import Data.Aeson.Types (parseEither)
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import Foreign
import Foreign.C.Types
import Foreign.C.String

foreign import ccall safe "stogas_verifier_abi_version" abi :: IO Word32
foreign import ccall safe "stogas_transport_start" start :: CString -> CSize -> Ptr (Ptr ()) -> IO CString
foreign import ccall safe "stogas_transport_close" close :: Ptr () -> IO ()
foreign import ccall safe "stogas_transport_free" freeTransport :: Ptr () -> IO ()
foreign import ccall safe "stogas_verifier_string_free" freeString :: CString -> IO ()

withTransport :: Value -> (String -> IO a) -> IO a
withTransport configuration action = bracket acquire release (action . snd)
  where
    acquire = do
      version <- abi
      unless (version == 1) $ fail "Unsupported verifier ABI"
      B.useAsCStringLen (L.toStrict $ encode configuration) $ \(bytes, size) ->
        alloca $ \output -> do
          poke output nullPtr
          raw <- start bytes (fromIntegral size) output
          handle <- peek output
          (do
            result <- bracket (pure raw) freeString $ \value -> do
              unless (value /= nullPtr) $ fail "Unable to start verified transport"
              B.packCString value
            url <- case eitherDecodeStrict result >>= parseEither envelope of
              Right value | handle /= nullPtr -> pure value
              _ -> fail "Unable to start verified transport"
            pure (handle, url)
            ) `onException` freeTransport handle
    release (handle, _) = close handle >> freeTransport handle
    envelope = withObject "transport" $ \value -> do
      ok <- value .: "ok"
      unless ok $ fail "Transport rejected configuration"
      result <- value .: "value"
      result .: "base_url"

File: examples/ocaml-http/transport.ml
open Ctypes

let with_transport library configuration action =
  let native = Dl.dlopen ~filename:library ~flags:[Dl.RTLD_NOW; Dl.RTLD_LOCAL] in
  let bind name signature = Foreign.foreign ~from:native name signature in
  let abi = bind "stogas_verifier_abi_version" (void @-> returning uint32_t) in
  let start = bind "stogas_transport_start"
    (string @-> size_t @-> ptr (ptr void) @-> returning (ptr char)) in
  let close = bind "stogas_transport_close" (ptr void @-> returning void) in
  let free = bind "stogas_transport_free" (ptr void @-> returning void) in
  let free_string = bind "stogas_verifier_string_free" (ptr char @-> returning void) in
  if Unsigned.UInt32.to_int (abi ()) <> 1 then failwith "Unsupported verifier ABI";
  let bytes = Yojson.Safe.to_string configuration in
  let output = allocate (ptr void) null in
  let raw = start bytes (Unsigned.Size_t.of_int (String.length bytes)) output in
  Fun.protect ~finally:(fun () -> close !@output; free !@output) (fun () ->
    let result = Fun.protect ~finally:(fun () -> free_string raw) (fun () ->
      if is_null raw then failwith "Unable to start verified transport";
      Yojson.Safe.from_string (coerce (ptr char) string raw)) in
    let open Yojson.Safe.Util in
    if is_null !@output || member "ok" result <> `Bool true then
      failwith "Unable to start verified transport";
    let url = result |> member "value" |> member "base_url" |> to_string in
    action url)
